client/src/utilities/solvingAlgorithms.ts

```diff
@@ -0,0 +1,233 @@
+            //   if (curr.x === start.x && curr.y === start.y && seen.flat().filter(val => val === true).length > 1) {
+            //     console.log("No path to endpoint found");
+            //     const deleteLastCurr = path.pop();
+            //     if (deleteLastCurr) {
+            //       const { x: deleteX, y: deleteY } = deleteLastCurr;
+            //       const updatedMazeBackTrack: number[][] = maze.map(
+            //         (row, rowIndex) =>
+            //           row.map((cell, colIndex) =>
+            //             rowIndex === deleteX && colIndex === deleteY
+            //         ? 0
+            //         : cell
+            //       )
+            //     );
+            //     setMaze(updatedMazeBackTrack);
+            //   }
+            //   resolve(true);
+            //   return;
+            // }
+
+            //* if seen the spot already (using argument of seen that is boolean 2D array)
+            if (seen[curr.x][curr.y]) {
+                resolve(false);
+                return;
+            }
+
+            //*2. recurse steps
+            //* pre
+            seen[curr.x][curr.y] = true;
+            path.push(curr);
+            console.log(seen);
+            console.log(path);
+
+            const updatedMaze: number[][] = maze.map((row, rowIndex) =>
+                row.map((cell, colIndex) =>
+                    rowIndex === curr.x && colIndex === curr.y ? 2 : cell
+                )
+            );
+            setMaze(updatedMaze);
+
+            //* recurse
+
+            for (const direction of directions) {
+                if (
+                    await dfs(
+                        updatedMaze,
+                        direction,
+                        start,
+                        end,
+                        seen,
+                        path,
+                        setMaze,
+                        delay,
+                        setSolving,
+                        solvingRef
+                    )
+                ) {
+                    resolve(true);
+                    return;
+                }
+            }
+
+            //* post
+            const deleteCurr = path.pop();
+            console.log(`popped x:${deleteCurr.x} , y:${deleteCurr.y}`);
+            console.log(path);
+
+            if (deleteCurr) {
+                const { x: deleteX, y: deleteY } = deleteCurr;
+                const updatedMazeBackTrack: number[][] = maze.map(
+                    (row, rowIndex) =>
+                        row.map((cell, colIndex) =>
+                            rowIndex === deleteX && colIndex === deleteY
+                                ? 0
+                                : cell
+                        )
+                );
+                // console.log(updatedMazeBackTrack);
+                setMaze(updatedMazeBackTrack);
+            }
+
+            resolve(false);
+            return;
+        }, delay);
+    });
+};
```

Comment:
Consider this refactor as reference on how you can clean up the code and make it more readable:

```
import { Point } from "./types";

export const dfs = async (
    maze: number[][],
    curr: Point,
    end: Point,
    seen: boolean[][],
    path: Point[],
    setMaze: React.Dispatch<React.SetStateAction<number[][]>>,
    delay: number,
    setSolving: React.Dispatch<React.SetStateAction<boolean>>,
    solvingRef: React.MutableRefObject<boolean>
): Promise<boolean> => {
    // Check if the solving process has been stopped
    if (!solvingRef.current) {
        console.log("Stopped solving");
        setSolving(false);
        return false;
    }

    // Base case: Check if the current position is out of bounds
    if (curr.x < 0 || curr.x >= maze.length || curr.y < 0 || curr.y >= maze[0].length) {
        return false;
    }

    // Base case: Check if the current cell is a wall or already seen
    if (maze[curr.x][curr.y] === 1 || seen[curr.x][curr.y]) {
        return false;
    }

    // Base case: Check if the current position is the end point
    if (curr.x === end.x && curr.y === end.y) {
        path.push(end);
        updateMaze(maze, curr, setMaze, 2);
        setSolving(false);
        return true;
    }

    // Mark the current cell as seen and add to path
    seen[curr.x][curr.y] = true;
    path.push(curr);
    updateMaze(maze, curr, setMaze, 2);

    // Introduce a delay for visualization purposes
    await new Promise((resolve) => setTimeout(resolve, delay));

    // Define possible directions for DFS (right, up, left, down)
    const directions: Point[] = [
        { x: curr.x + 1, y: curr.y },
        { x: curr.x, y: curr.y - 1 },
        { x: curr.x - 1, y: curr.y },
        { x: curr.x, y: curr.y + 1 }
    ];

    // Recursively explore each direction
    for (const direction of directions) {
        if (await dfs(maze, direction, end, seen, path, setMaze, delay, setSolving, solvingRef)) {
            return true;
        }
    }

    // Backtrack: Remove the current cell from the path and reset it in the maze
    const deleteCurr = path.pop();
    if (deleteCurr) {
        updateMaze(maze, deleteCurr, setMaze, 0);
    }

    return false;
};

// Function to update the maze state
const updateMaze = (
    maze: number[][],
    point: Point,
    setMaze: React.Dispatch<React.SetStateAction<number[][]>>,
    value: number
) => {
    const updatedMaze = maze.map((row, rowIndex) =>
        row.map((cell, colIndex) =>
            rowIndex === point.x && colIndex === point.y ? value : cell
        )
    );
    setMaze(updatedMaze);
};
```